src/backend/core/warren/base_indicator.py

```diff
@@ -2,9 +2,75 @@
 from abc import ABC, abstractmethod
 from typing import List
 
+import numpy as np
+import pandas as pd
 from ralph.backends.http.lrs import LRSQuery
 
 
+def parse_raw_statements(raw_statements) -> pd.DataFrame:
+    """Parse LRS statements, explode the columns, and add a `date` column."""
+    flattened = pd.json_normalize(raw_statements)
+    # Disable chained assignment warning to make the transformation inplace
+    pd.options.mode.chained_assignment = None
+    # Transform timestamp column into a date with day (YYYY-MM-DD)
+    flattened.loc[:, "date"] = pd.to_datetime(flattened.loc[:, "timestamp"]).dt.date
+    return flattened
+
+
+def add_uniformized_actor_id(statements: pd.DataFrame) -> pd.DataFrame:
+    """Add a `unique_actor_id` column that uniquely identifies the agent.
+
+    Depending on the xAPI statements, the actor can be identified in 4 different ways :
+    https://github.com/adlnet/xAPI-Spec/blob/master/xAPI-Data.md#details-4. This
+    function handles the 4 cases and creates a `unique_actor_id` column that
+    can be used later without worrying about the 4 IFIs.
+    """
+    COLUMNS_TO_CHECK = {
+        "actor.account.name",
+        "actor.account.homePage",
+        "actor.mbox",
+        "actor.mbox_sha1sum",
+        "actor.openid",
+    }
+
+    for column in COLUMNS_TO_CHECK - set(statements.columns):
+        statements[column] = np.nan
+
+    def generate_unique_agent_id(row):
+        non_nan_columns = {col for col in COLUMNS_TO_CHECK if pd.notna(row[col])}
+        if len(non_nan_columns) == 0:
+            raise ValueError(
+                "No way of identifying the agent: all identification columns are NaN."
+            )
+        elif len(non_nan_columns) == 1:
+            return row[non_nan_columns.pop()]
+        elif len(non_nan_columns) > 1:
+            # If there are 2 identification columns, it must be both
+            # `actor.account.name` and `actor.account.homePage` to be valid
+            if non_nan_columns == {"actor.account.name", "actor.account.homePage"}:
+                return f"{row['actor.account.name']}-{row['actor.account.homePage']}"
+            else:
+                raise ValueError(
+                    "Multiple ways of identification for agent: "
+                    "more than one identification column is filled."
+                )
+        else:
+            raise ValueError("Unexpected number of identification columns.")
+
+    statements["actor.unique_actor_id"] = statements.apply(
+        generate_unique_agent_id, axis=1
+    )
```

Comment:
I haven't tested it, but WDYT about the following alternative implementation?

```suggestion
    xapi_actor_identifier_columns = XAPI_ACTOR_IDENTIFIER_PATHS.intersection(set(statements.columns))
    
    if len(xapi_actor_identifier_columns) == 0:
        raise ValueError(
            "There is no way of identifying the agent in submitted statements."
        )

    statements["actor.uid"] = statements.apply(
        lambda row, columns: hashlib.sha256("-".join(row[columns]).hexdigest(), 
        axis=1, 
        xapi_actor_identifier_columns
    )
```

Reply:
Neat 🔥